app/models/submission.rb

```diff
@@ -0,0 +1,40 @@
+class Submission < ApplicationRecord
+  belongs_to :user
+
+  has_many :comments, class_name: "Submission", foreign_key: :post_id
+  belongs_to :post, class_name: "Submission", optional: true
+
+  validates :user, presence: true
+  validates :title, :url, presence: true, if: :url_submission?
+
+  validate :text_submission
+  validate :at_least_one_field_present
+
+  def limit_text
+    ary = self.text.chars.shift(80)
+    shift_pos = 0
+    ary.reverse.each_with_index do |val, index|
+      shift_pos = index + 1 if val.blank?
+      break if shift_pos > 0
+    end
+    "#{ary.shift(ary.count - shift_pos).join}..."
+  end
+
+  private
+
+    def url_submission?
+      url.present? || title.present?
+    end
+
+    def text_submission
+      if text.present? && (url.present? || title.present?)
```

Comment:
No action required: I just learned that `.present?` returns `false` on sparsely empty strings (like `'   '`). Neat.

Reply:
I learned that from Pat last week.

Reply:
Can use the `url_submission?` method here?

Reply:
Technically, yes, but I feel like that would make the code less human readable.

Actually, I don't like how I currently define the 2 submission types - with one being a boolean and the other adding an error message. It seems to obscure the either one or the other relationship.

Reply:
Following my last comment - here's a refactor https://github.com/gingin77/hackernews_clone/pull/5/commits/009139844e27e461bcdd49c4af25fd92a2a494ea